wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,866 @@
+      ]
+   ```
+
+For the majority of estimators, the above steps will be sufficient to correctly work with the cuML library and ensure a consistent API. However, situations may arrise where an estimator differs from the standard pattern and some of the functionality needs to be customized. The remainder of this guide takes a deep dive into the estimator functionality to assist developers when building estimators.
+
+## Background
+
+Some background is necessary to understand the design of estimators and how to work around any non-standard situations.
+
+### Input and Output Types in cuML
+
+In cuML we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:
+
+ - cuDF DataFrame or Series
+ - Pandas DataFrame or Series
+ - NumPy Arrays
+ - Numba Device Arrays
+ - CuPy arrays
+ - CumlArray type (Internal to the `cuml` API only.)
+
+When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.
+
+Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty, though they may still increase memory usage.
+
+Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.
+
+Internally, all arrays should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.
+
+### Speficying the Array Output Type
+
+Users can choose which array type should be returned by cuml by either: 
+1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)
+2. Globally setting the `cuml.global_output_type`
+3. Temporarily setting the `cuml.global_output_type` via the `cuml.using_output_type` context manager
+
+**Note:** Setting the global output type <u>will take precedence over any value in `Base.output_type`</u>
+
+Changing the array output type will alter the return value of estimator functions (i.e. `predict()`, `transform()`), and the return value for array-like estimator attributes (i.e. `my_estimator.classes_` or `my_estimator.coef_`)
+
+All output_types (including `cuml.global_output_type`) are specfied using an all lowercase string. These strings can be passed in an estimators constructor or via `cuml.set_global_output_type` and `cuml.using_output_type`. Accepted values are:
+
+ - `None`: (Default) No global value set. Will use individual values from estimators output_type
+ - `"input"`: Similar to `None`. Will mirror the same type as any array passed into the estimator
+ - `"mirror"`: Special value for handling internal `cuml` API calls. Will mirror the input value set directly
```

Comment:
what is the difference in behavior between these 2? If `mirror` is only for internal API calls, I would put it in the end separated, stating that it is for developer use only while the rest are user accessible

Reply:
Moved mirror to a note at the end to indicate how special it is